### PR TITLE
[PLATFORM-1008]: Use macros for configuration

### PR DESCRIPTION
**Pull request by cpiemontese**
https://prima-assicurazioni-spa.myjetbrains.com/youtrack/issue/PLATFORM-1008

This PR set out to try to use macros to improve configuration, à la bridge_ex. After thinking about it I couldn't come up with a valid use case for macros, since Telepoison is used as a "static" library and not to define clients, as is the case for bridge_ex.

Transforming it to a macro would require people to `use` it somewhere and then access injected functions, which would probably be more verbose than anything.

I then tried to improve and fix what I could, so this PR includes the following changes:

- configuration options are correctly validated during `setup` instead of upon retrieval, as suggested by @mbusi 
- request options (e.g. `:ot_resource_route` and `:ot_attributes`) can be correctly passed together to Telepoison calls (e.g. `get`)
- non-valid `:ot_attributes` are filtered out instead of being mapped to `nil`
- usages of `:infer_fn` are now all `:infer_route`
- moved config to a `Configuration` module, which is also a struct (used internally for now)

### CODE REVIEW

File: lib/configuration.ex [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+  end
+
+  @doc """
+  Get a configuration value if present and `setup` has been called or return an error
+  """
+  @spec get(:infer_fn | :ot_attributes) :: {:ok, any()} | {:error, String.t()}
+  def get(key)
+
+  def get(:infer_fn) do
+    if agent_started?() do
```
Comment (mbusi) [inline lib/configuration.ex:59]: I don't like the defensive approach that much (making a `Process.whereis` every time you call this function). Can't you simply break? the agent is supposed to be always there if your client correctly called the `setup/1` funciton, right? If you didn't want the process to exit you could simply add a catch clause to your functions like so:
```elixir
catch
    :exit, {:noproc, _} -> {:error, "Route inference function hasn't been configured"}
end
```

File: lib/configuration.ex [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+      Agent.get(
+        __MODULE__,
+        fn
+          {infer_fn, _} when is_function(infer_fn, 1) ->
```
Comment (mbusi) [inline lib/configuration.ex:63]: shouldn't you do these kind of configuration validations during setup (fail early)?

**Comment by cpiemontese** [comment]
I converted this PR to a draft because it's a WIP "redesign" of how we configure Telepoison and I'm still poking around to understand how to proceed

The `setup` function is a bit misleading because it's actually optional and should probably be `configure` or something along those lines

Since it's implemented with an `Agent` the agent may not be there and that's why we check if the `Agent` has been started, but I think your suggestion of catching is better. As far as validating earlier: ideally yes, I'll try to see if I can change it to make it so.

In general, we were wondering if there was a way to make Telepoison more like `bridge_ex` and pass configuration using a macro, but I don't even know if it makes sense for this use case

**Review by MaeIsBad — COMMENTED** [review]
2 relatively minor questions

File: lib/configuration.ex [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+  @default_route_inference_function &Telepoison.URI.infer_route_from_request/1
+
+  defstruct [:infer_route, :ot_attributes]
+
+  alias HTTPoison.Request
+
+  @spec setup(infer_route: (Request.t() -> String.t()), ot_attributes: [{String.t(), String.t()}]) :: :ok
+  def setup(opts \\ []) do
+    Agent.start_link(fn -> set_configuration(opts) end, name: __MODULE__)
```
[the remark this answers is not in the document]
Comment (mbusi) [reply lib/configuration.ex:12]: > Are we sure that using an Agent here is the best way to do this? Wouldn't an application env keyed by the module name make more sense?

This is honestly what I wanted to originally suggest, but didn't mean to be too rude :laughing: 
It seems it's the more 'erlang' way to do it tbh

[the remark this answers is not in the document]
Comment (cpiemontese) [reply lib/configuration.ex:12]: I completely forgot about that pattern :facepalm: Yes, it would be more natural, I'll implement it that way

File: lib/configuration.ex [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        # Unset, return default function
+        nil ->
+          @default_route_inference_function
+
+        infer_fn when is_function(infer_fn, 1) ->
+          infer_fn
+
+        # Set incorrectly
+        _ ->
+          raise RuntimeError, "The configured :infer_route keyword option value must be a function with an arity of 1"
```
[the remark this answers is not in the document]
Comment (cpiemontese) [reply lib/configuration.ex:29]: @MaeIsBad I based my decision mainly on [this](https://elixirforum.com/t/would-you-raise-this-as-an-argumenterror-runtimeerror-or-other/26050) post because I was thinking about configuration as envs.

I now think `ArgumentError` would be more appropriate since this is in response to a call to `setup`, but if we change it to be `Application`-based I think `RuntimeError` becomes more appropriate again...

TL;DR I'll think about it again while refactoring to read the env

File: lib/configuration.ex (hunk 3 above)
[the remark this answers is not in the document]
Comment (mbusi) [reply lib/configuration.ex:12]: https://hexdocs.pm/elixir/1.13/Application.html#c:start/2 <- this is were you should look at (IMO) for this kind of stuff; if it's an application (doing things at startup) we should treat it as such. This is what I meant with the other comment, that, to be honest, is poorly written -.-"
Talking about
> this is an application (of sort, more like a shallow application used to start and configure a collection of applications)

[the remark this answers is not in the document]
Comment (mbusi) [reply lib/configuration.ex:12]: More on that: as far as I remember elixir adds every dep to the required applications list when generating an erlang release so you wouldn't even need to actively call `Telepoison.setup` if you'd implement a proper `Application.start/2` function. Can't remember why it wasn't implemented like that tbh :shrug:

[the remark this answers is not in the document]
Comment (cpiemontese) [reply lib/configuration.ex:12]: I think it was implemented like this because `Telepoison` technically doesn't need to be set up, it's just a wrapper around HTTPoison. Also `setup` would be better called `set_default_options` or `configure` or something like that

If I understand correctly there are (at least) four ways to handle "configuring" something in Elixir

1. Define a macro that just substitutes those values e.g. bridge_ex -> not applicable here
2. Fetch stuff from `Application` env -> possibly risky if `Telepoison` is used in umbrella apps and each app wants their own config
3. Use an `Agent` like we are doing here -> I think it incurs the same problem from **2** since the agent would have the same name for each umbrella app
4. Leverage `Application.start` to define startup values -> `Telepoison` shouldn't need to be started up this way because it should be just a wrapper

Did I understand correctly?

[the remark this answers is not in the document]
Comment (mbusi) [reply lib/configuration.ex:12]: > Telepoison shouldn't need to be started up this way because it should be just a wrapper

Funny thing is elixir will do it anyway, as it does for every dependency you add; they get added to the erlang release as the main application dependencies, so they get started before your app anyway

> possibly risky if Telepoison is used in umbrella apps and each app wants their own config

I think that is true for every approach you mentioned and it is not how umbrella apps work anyway (they shared the conf anyway); point 4 can be implemented via point 2 or 3 btw :D

[the remark this answers is not in the document]
Comment (cpiemontese) [reply lib/configuration.ex:12]: Ok! I'll try to do 2 and 4 then :smile:

File: test/support/test_api.ex [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+    Application.put_env(:telepoison, key, value)
+
+    on_exit(fn ->
+      case env_to_restore do
+        :unset -> Application.delete_env(:telepoison, key)
+        _ -> Application.put_env(:telepoison, key, env_to_restore)
+      end
+    end)
+  end
+end
```
Comment (MaeIsBad) [inline test/support/test_api.ex]: I think instead we could do:

```suggestion
defmodule Telepoison.TestCase do
  use ExUnit.CaseTemplate, async: false
  
  setup do
    env_to_restore = Application.get_env_all(:telepoison)

    on_exit(fn ->
        _ -> Application.put_all_env([{:telepoison, env_to_restore}])
      end
    end)
  end
end
```

and then `use Telepoison.TestCase` instead of `ExUnit.Case`

**Review by MaeIsBad — COMMENTED** [review]
Looks mostly fine. Do we need to add new docs about how to configure telepoison?

File: test/support/test_api.ex (hunk 5 above)
Comment (MaeIsBad) [reply test/support/test_api.ex]: After testing I relalized that this is broken because it doesn't remove keys that didn't exist before, only overwrites them. This should work

```suggestion
defmodule Telepoison.TestCase do
  use ExUnit.CaseTemplate, async: false

  setup do
    env_to_restore = Application.get_all_env(:telepoison)

    on_exit(fn ->
      Application.get_all_env(:telepoison)
      |> Keyword.keys()
      |> Enum.map(fn k -> Application.delete_env(:telepoison, k) end)

      Application.put_all_env([{:telepoison, env_to_restore}])
    end)
  end
end
```

Comment (cpiemontese) [reply test/support/test_api.ex]: Since there are only two keys I just added two deletes for now but yes, `put_all_env` with an empty list doesn't do the job :disappointed:

**Review by MaeIsBad — DISMISSED** [review]

**Review by MaeIsBad — APPROVED** [review]

File: README.md [hunk 6]
```diff
@@ -24,22 +24,21 @@ Telepoison.get!(url, headers, opts)
 
 ## Configuration
 
-`Telepoison.setup/1` takes a `Keyword list` that can configure:
+Telepoison can be configured through `config :telepoison` or using `Telepoison.setup/1` which is deprecated. The configurable options are:
```
Comment (MaeIsBad) [inline README.md:27]: I would just remove the mention of the deprecated function completely

**Review by MaeIsBad — APPROVED** [review]
